src/net_processing.cpp

```diff
@@ -889,15 +889,24 @@ void PeerManager::InitializeNode(CNode *pnode) {
 
 void PeerManager::ReattemptInitialBroadcast(CScheduler& scheduler) const
 {
-    std::map<uint256, uint256> unbroadcast_txids = m_mempool.GetUnbroadcastTxs();
-
-    for (const auto& elem : unbroadcast_txids) {
-        // Sanity check: all unbroadcast txns should exist in the mempool
-        if (m_mempool.exists(elem.first)) {
-            LOCK(cs_main);
+    std::vector<std::pair<uint256, uint256>> relay_transactions;
+    {
+        LOCK(m_mempool.cs);
+        std::map<uint256, uint256> unbroadcast_txids = m_mempool.GetUnbroadcastTxs();
+        relay_transactions.reserve(unbroadcast_txids.size());
+        for (const auto& elem : unbroadcast_txids) {
+            // Sanity check: all unbroadcast txns should exist in the mempool
+            if (m_mempool.exists(elem.first)) {
+                relay_transactions.push_back(elem);
+            } else {
+                m_mempool.RemoveUnbroadcastTx(elem.first, true);
+            }
+        }
+    }
+    {
+        LOCK(cs_main);
+        for (const auto& elem : relay_transactions) {
             RelayTransaction(elem.first, elem.second, m_connman);
-        } else {
-            m_mempool.RemoveUnbroadcastTx(elem.first, true);
         }
     }
 
```

Comment:
This can be simplified to:

```cpp
void PeerManager::ReattemptInitialBroadcast(CScheduler& scheduler) const
{
    for (const auto& elem : WITH_LOCK(m_mempool.cs, return m_mempool.GetUnbroadcastTxs())) {
        LOCK(cs_main);
        RelayTransaction(elem.first, elem.second, m_connman);
    } 

    // Schedule next run for 10-15 minutes in the future.
    ...
}
```

Because `m_mempool.exists()` will always return `true` for a tx returned by `m_mempool.GetUnbroadcastTxs()` if we don't release `m_mempool.cs` between the two calls.

Also, the tx could be removed after we release `m_mempool.cs` and before we call `RelayTransaction()` and this is ok and is [handled just fine](https://github.com/bitcoin/bitcoin/blob/564e1ab/src/net_processing.cpp#L4429).

cc @gzhao408

Reply:
But is it really necessary to have `cs_main` lock in each iteration?

Reply:
Since it was locked in each iteration before this PR, the question should rather be "If we move `LOCK(cs_main)` before the loop, why would we do that?"

The frequent lock/unlock allows other threads to proceed. I don't see a reason to change it.

Reply:
> The frequent lock/unlock allows other threads to proceed. I don't see a reason to change it.

Agree (https://github.com/bitcoin/bitcoin/pull/19917#discussion_r485047637).

Reply:
@vasild I don't agree with that. Other threads can proceed but the current thread will wait unnecessarily in each iteration for the lock and as such other things will be delayed, not mentioning the mutex overhead. See https://stackoverflow.com/a/3652428.

In this case `RelayTransaction` is pretty quick, nothing that can potentially cause a big lock on `cs_main`.